NOTES/visionmaker/xinput-1.2/xinput.man
.\" $Id: xinput.man,v 1.1.1.1 2005/10/09 04:28:50 stanonik Exp $
.TH xinput 1 "$Date: 2005/10/09 04:28:50 $" "Frederic Lepied"

.SH NAME
xinput - utility to configure and test XInput devices

.SH SYNOPSIS
.B xinput
[version] [list [\fIdevice_name\fP]] [set-pointer \fIdevice_name\fP]
[get-feedbacks \fIdevice_name\fP]
[set-mode \fIdevice_name\fP \fIABSOLUTE|RELATIVE\fP]
[set-ptr-feedback \fIdevice_name\fP \fIthreshold\fP \fInum\fP \fIdenom\fP]
[set-integer-feedback \fIdevice_name\fP \fIindex\fP \fIvalue\fP]
[set-button-map \fIdevice_name\fP \fImap button 1\fP [\fImap button 2\fP [\fI...\fP]]]
[query-state \fIdevice_name\fP]
[test [-proximity] \fIdevice_name\fP]

.SH DESCRIPTION
.TP 8
.B xinput version
test if the XInput extension is available and return the version number
of the program.
.PP
.TP 8
.B xinput list [\fIdevice_name\fP]
If no argument is given list all the input devices showing all their
features. If an argument is given, show all the feature of \fIdevice_name\fP.
Uses XListInputDevices(3).
.PP
.TP 8
.B xinput get-feedbacks \fIdevice_name\fP
Display the feedbacks of \fIdevice_name\fP. Uses XGetFeedbackControl(3).
.PP
.TP 8
.B xinput set-pointer \fIdevice_name\fP
Switch \fIdevice_name\fP in core pointer. Uses XChangePointerDevice(3).
.PP
.TP 8
.B xinput set-mode \fIdevice_name\fP \fIABSOLUTE|RELATIVE\fP
Change the mode of \fIdevice_name\fP. Uses XSetDeviceMode(3).
.PP
.TP 8
.B xinput set-ptr-feedback \fIdevice_name\fP \fIthreshold\fP \fInum\fP \fIdenom\fP
Change the acceleration of \fIdevice_name\fP. Uses XChangeFeedbackControl(3).
.PP
.TP 8
.B xinput set-integer-feedback \fIdevice_name\fP \fIindex\fP \fIvalue\fP
Change the value of an integer feedback of \fIdevice_name\fP. Uses XChangeFeedbackControl(3).
.PP
.TP 8
.B xinput set-button-map \fIdevice_name\fP \fImap button 1\fP [\fImap button 2\fP [\fI...\fP]]
Change the button mapping of \fIdevice_name\fP. Uses XSetDeviceButtonMapping(3).
.PP
.TP 8
.B xinput query-state \fIdevice_name\fP
Query the device state. Uses XQueryDeviceState(3).
.PP
.TP 8
.B xinput test [-proximity] \fIdevice_name\fP
Register all extended events from \fIdevice_name\fP and enter an endless
loop displaying events received. If the -proximity is given, ProximityIn
and ProximityOut are registered.
.PP
\fIdevice_name\fP can be the device name as a string or the XID of the
device.
.PP

.SH COPYRIGHT
Copyright 1996,1997, Frederic Lepied.

.SH AUTHOR

.nf
Frederic Lepied, France <Frederic.Lepied@sugix.frmug.org>
.fi

Patches, bug reports, and suggestions are welcome.


pad/ChangeLog
=========================
Version 0.9b - 6/25/97

  PADDRAW APPLICATION CHANGES
  ---------------------------
  * Changed default zooming behavior.  Now, when you move the
    cursor while you zoom, the surface follows you as if you
    were panning.  Same with zooming selected objects.  This approach
    seems to be slightly less powerful, but is much easier to learn
    as it is consistent with panning, and doesn't have "backwards"
    behavior.  You can use the old zooming style if you want by
    turning of "Zoom follows cursor" on the preferences panel.
    Thanks to Ron Hightower for figuring out this one!

  * Added start-up warning if running Pad++ in > 8 bit mode.

  * Added (hopefully) helpful startup tips.

  * Reshape handles now allow squashing and stretching of all objects
    except text and aliases.
  
  PADDRAW BUG FIXES
  -----------------
	
  TCL INTERFACE CHANGES
  ---------------------
  * INCOMPATIBLE CHANGE: Changed implementation of -width and -height
    itemconfigure options.  They still set the width and height of objects,
    but now they consistently squash/stretch the object (whenever possible)
    instead of clipping it.  The only object for which this is not true
    is text and aliases which is still clipped since we do not have the 
    capability to squash/stretch it.

    This change was necessary to make -width/-height consistent.  For
    some objects (all the widgets and images), this is what the behavior
    already was.  If there is a strong demand for the old clipping 
    behavior, I will consider adding it back with -clipwidth/-clipheight
    options.  Note that for groups, clipping behavior is available with
    a panel object.

    Specifically, the object types for which this behavior changed is:
	portal
	line
	Spline
	Polygon
	Rectangle
	Oval
        Group
        Grid

  * INCOMPATIBLE CHANGE: Changed -dim parameter format of new 'grab'
    command to be -dim {width height} instead of -dim {<width>x<height>}
    to be more Tcl'ish.

  * INCOMPATIBLE CHANGE: Added 'render draw filledborder' command, and
    changed 'render draw border' to now draw an unfilled border.

  * Added 'animate' command. The animate command allows asynchronous
    animations of most options of items on the Pad++ surface.  An item can
    be moved across the screen while its color is being changed while
    another is being rotated.  This all happens in the background so that
    Pad++ continues to process events while animations happen. 

  * Added new item types
	checkbox   w/ same itemconfigure options as button
	menubar    w/ itemconfigure options -fill -members
	menu       w/ itemconfigure options as button and -members -memberlabels
	choicemenu w/ itemconfigure options as menu but witthout -text
	menuitem   w/ itemconfigure options as button
	checkboxmenuitem w/ itemconfigure options as button
	Added the -menubar option to frame  

  * Made 'text' command work on textarea items

  C++ SUBSTRATE BUG FIXES
  -----------------------

  * Made it so Tcl errors in button and scrollbar callbacks properly
    generate Tcl background errors.

  	* Fixed bug where zoomactions didn't fire on transparent objects.
	
  * Detected if random() method exists.  If not, use rand().

  * Fixed bug where if you clicked on an object, moved off it, and released,
    after the release, a B1-Leave got generated instead of a Leave.  This
    used to work in pad-0.2.7, but was broken in pad-0.9b.

  * Fixed portals so visiblelayers works properly when looking on to another surface.

  * Fixed many Windows memory bugs and cleaned up distribution so
    it now compiles cleanly under Windows NT and 95, and seems
    to work pretty well.

  * Added Windows binaries to distribution.

  * Fixed 'make install'.  It died because the "Register" file
    is gone.

  * Fixed bug where changing the layer of a grid would cause it to move.
    Now it doesn't.

  * Fixed 'grab' command to work with 24 bit visuals - even taking
    into account different byte orders of X servers, etc.
   	
  * Hooked up vertical scrollbars to textarea widget.
	
  * Fixed bug with user-defined types where parameters before options
    didn't get passed on to the Tcl procedure properly.

  * Fixed a nasty portal/event bug where motion events in the portal would
    sometimes generate bad (%i, %j) coordinates.  This had the side effect
    that zooming within a portal would sometimes jump around.

  * Fixed another portal/event bug where events didn't go through
    portals that were members of groups properly.

  * Fixed another portal/event bug where events didn't always
    get generated when going through a portal to another pad surface.

  * Fixed configure bug where the architecture wasn't set right
    on 64 bit IRIX machines.  Now, the bin directory is
    bin-MIPS-IRIX for all IRIX architectures.

  * Fixed bug where 'center' command did not work properly when
    centering within portals and the portal had been scaled.
	
=========================
Version 0.9b - 4/27/97

  SUMMARY OF IMPORTANT CHANGES
  ----------------------------

  * INCOMPATIBLE CHANGE:  Changed usage of hastag command
    to be compatible with addtag and deletetag commands.  New usage is:
       .pad hastag tagToCheck tagOrId
    The order of the parameters used to be backward.

  * INCOMPATIBLE CHANGE: Sticky Objects Changed/Extended
    Reimplemented sticky objects based on constraints 
    instead of a separate display list.  Now, sticky objects are identical
    in every way to regular objects except that are kept on a special list.  
    Every time the top-level view is changed, the position and size of sticky
    objects are recomputed to behave properly.  Several new types of
    sticky objects are now available, including objects that are sticky 
    just in Z (but not in x, y), and "sticky view" objects that always
    stay within the view.  See the ChangeLog or reference documentation

NOTES/visionmaker/xinput-1.2/xinput.c
/*
 * Copyright 1996 by Frederic Lepied, France. <Frederic.Lepied@sugix.frmug.org>
 *                                                                            
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is  hereby granted without fee, provided that
 * the  above copyright   notice appear  in   all  copies and  that both  that
 * copyright  notice   and   this  permission   notice  appear  in  supporting
 * documentation, and that   the  name of  Frederic   Lepied not  be  used  in
 * advertising or publicity pertaining to distribution of the software without
 * specific,  written      prior  permission.     Frederic  Lepied   makes  no
 * representations about the suitability of this software for any purpose.  It
 * is provided "as is" without express or implied warranty.                   
 *                                                                            
 * FREDERIC  LEPIED DISCLAIMS ALL   WARRANTIES WITH REGARD  TO  THIS SOFTWARE,
 * INCLUDING ALL IMPLIED   WARRANTIES OF MERCHANTABILITY  AND   FITNESS, IN NO
 * EVENT  SHALL FREDERIC  LEPIED BE   LIABLE   FOR ANY  SPECIAL, INDIRECT   OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA  OR PROFITS, WHETHER  IN  AN ACTION OF  CONTRACT,  NEGLIGENCE OR OTHER
 * TORTIOUS  ACTION, ARISING    OUT OF OR   IN  CONNECTION  WITH THE USE    OR
 * PERFORMANCE OF THIS SOFTWARE.
 *
 */

#include "xinput.h"
#include <ctype.h>

static const char rcs_id[] = "$Id: xinput.c,v 1.1.1.1 2005/10/09 04:28:50 stanonik Exp $";

typedef int (*prog)(
#if NeedFunctionPrototypes
		    Display* display, int argc, char *argv[],
		    char *prog_name, char *prog_desc
#endif
);

typedef struct 
{
    char	*func_name;
    char	*arg_desc;
    prog	func;
} entry;

static entry drivers[] =
{
    {"get-feedbacks",
     "<device name>",
     get_feedbacks
    },
    {"set-ptr-feedback",
     "<device name> <threshold> <num> <denom>",
     set_ptr_feedback
    },
    {"set-integer-feedback",
     "<device name> <feedback id> <value>",
     set_integer_feedback
    },
    {"set-button-map",
     "<device name> <map button 1> [<map button 2> [...]]",
     set_button_map
    },
    {"set-pointer",
     "<device name> [<x index> <y index>]",
     set_pointer
    },
    {"set-mode",
     "<device name> ABSOLUTE|RELATIVE",
     set_mode
    },
    {"list",
     "[<device name>...]",
     list
    },
    {"query-state",
     "<device name>",
     query_state
    },
    {"test",
     "[-proximity] <device name>",
     test
    },
    {"version",
     "",
     version
    },
    {0, 0, 0
    }
};

static Bool
is_xinput_present(Display	*display)
{
    XExtensionVersion	*version;
    Bool		present;
    
    version = XGetExtensionVersion(display, INAME);

    if (version && (version != (XExtensionVersion*) NoSuchExtension)) {
	present = version->present;
	XFree(version);
	return present;
    } else {
	return False;
    }
}

XDeviceInfo*
find_device_info(Display	*display,
		 char		*name,
		 Bool		only_extended)
{
    XDeviceInfo	*devices;
    int		loop;
    int		num_devices;
    int		len = strlen(name);
    Bool	is_id = True;
    XID		id;
    
    for(loop=0; loop<len; loop++) {
	if (!isdigit(name[loop])) {
	    is_id = False;
	    break;
	}
    }

    if (is_id) {
	id = atoi(name);
    }
    
    devices = XListInputDevices(display, &num_devices);

    for(loop=0; loop<num_devices; loop++) {
	if ((!only_extended || (devices[loop].use == IsXExtensionDevice)) &&
	    ((!is_id && strcmp(devices[loop].name, name) == 0) ||
	     (is_id && devices[loop].id == id))) {
	    return &devices[loop];
	}
    }
    return NULL;
}

static void
usage() 
{
    entry	*pdriver = drivers;
    
    fprintf(stderr, "usage :\n");
    
    while(pdriver->func_name) {
	fprintf(stderr, "\txinput %s %s\n", pdriver->func_name,
		pdriver->arg_desc);
	pdriver++;
    }
}

int
main(int argc, char * argv[])
{
    Display	*display;
    entry	*driver = drivers;
    
    if (argc < 2) {
	usage();
	return EXIT_FAILURE;
    }
    
    display = XOpenDisplay(NULL);
    
    if (display == NULL) {
	fprintf(stderr, "Unable to connect to X server\n");
	return EXIT_FAILURE;
    }

    if (!is_xinput_present(display)) {
	fprintf(stderr, "%s extension not available\n", INAME);
	return EXIT_FAILURE;
    }
    
    while(driver->func_name) {
	if (strcmp(driver->func_name, argv[1]) == 0) {
	    int	r = (*driver->func)(display, argc-2, argv+2,
				    driver->func_name, driver->arg_desc);
	    XFlush(display);
	    return r;
	}
	driver++;
    }
    
    usage();
    
    return EXIT_FAILURE;
}

/* end of xinput.c
 */
